### ISSUE TITLE
Cannot invoke project dependency command on a library

### ISSUE THREAD

**Issue by natemcmaster** [dotnet/cli#2645]
**Update:** the error is only on libraries, not standalone. 

~~It looks like https://github.com/dotnet/cli/pull/2617 broke PackagedCommandSpecFactory in tools on standalone apps.~~
## Steps to reproduce

Create a tool that invokes tries to execute a dependency command where projectContext is a netstandard **library**.

``` c#
                new ProjectDependenciesCommandFactory(
                        projectContext.TargetFramework,
                        Constants.DefaultConfiguration,
                        null,
                        null,
                        projectContext.ProjectDirectory
                    )
                    .Create("dotnet-my-tool", 
                            new []{"--dispatch", "--verbose"}, 
                            projectContext.TargetFramework, 
                            Constants.DefaultConfiguration)
```

Repro project: https://github.com/natemcmaster/dotnet-sample-tool/tree/issue2645. 
## Expected behavior

ProjectDependenciesCommandFactory should create and execute a command on the netstandard library.
## Actual behavior

On a library project

```
Running C:\Users\namc\dev\dotnet-sample-tool\.dotnet\shared\Microsoft.NETCore.App\1.0.0-rc2-3002476\corehost.exe --runtimeconfig C:\Users\namc\dev\dotnet-sample-tool\SampleLibrary\bin\Debug\netstandard1.5\SampleLibrary.runtimeconfig.json --depsfile C:\Users\namc\dev\dotnet-sample-tool\SampleLibrary\bin\Debug\netstandard1.5\SampleLibrary.deps.json --additionalprobingpath C:\Users\namc\.nuget\packages C:\Users\namc\.nuget\packages\dotnet-my-tool\1.0.0-tmp1461321718.66929\lib\netstandard1.5\dotnet-my-tool.dll --dispatch --verbose
Process ID: 55528
Could not resolve coreclr path
```
## Environment data

Microsoft.DotNet.Cli.Utils/1.0.0-rc2-002476

`dotnet --info` output:

```
.NET Command Line Tools (1.0.0-rc2-002476)

Product Information:
 Version:     1.0.0-rc2-002476
 Commit Sha:  1a8d393b9b

```

Fails on windows and linux.

Trace: [corehost_trace.log.txt](https://github.com/dotnet/cli/files/232360/corehost_trace.log.txt)

cc @schellap @piotrpMSFT

**Comment by natemcmaster** [comment]
Updated the original issue description. It looks like I had two bugs conflated, so I redacted the description. The first bug was resolved by deleting muxer. The second bug, however, "Could not resolve coreclr path" error persists.

Incidentally, why don't the install scripts delete the muxer for in-place upgrades?

**Comment by schellap** [comment]
I took a first stab and this file doesn't exist:

`C:\Users\namc\dev\dotnet-sample-tool\SampleLibrary\bin\Debug\netstandard1.5\SampleLibrary.runtimeconfig.json`

probably because there is no target framework (`Microsoft.NETCore.App`) in the `project.json`.

**Comment by natemcmaster** [comment]
@schellap and I took a look at this offline. This looks like a design issue, not a bug. "Could not find coreclr" is accurate because there is no coreclr referenced from a library project.

This is a common scenario we want to enable for Entity Framework. User's often use EF on a library that is shared between web, desktop, etc.
```json
{
  "version": "1.0.0-*",
  "name": "Contoso.DataAccess",
  "dependencies": {
    "NETStandard.Library": "1.5.0-*",
    "Microsoft.EntityFrameworkCore.Sqlite": "1.0.0-*",
    "dotnet-ef": {
      "version": "1.0.0-*",
      "type": "build"
    }
  },
  "frameworks": {
    "netstandard1.5": {
      "imports": "portable-net452+win81"
    }
  },
  "tools": {
    "dotnet-ef": "1.0.0-*"
  }
}
```

How can a user run a dependency tool on this kind of project?

cc @DamianEdwards @rowanmiller @piotrpMSFT

**Comment by natemcmaster** [comment]
User's can workaround this by invoking the tool a different project, such as a "netcoreapp1.0" web project that references the library. This is unfortunately misleading because means "tools" can still install in a library.

Example: to add migrations to "Contoso.DataAccess", you need to invoke dotnet-ef on "Contoso.Web" which references "Contoso.DataAccess". 

I spoke offline with @rowanmiller and we are okay with this limitation for RC2, so I'm removing the blocking label and milestone.

Regardless, it would still be useful to invoke "dependency"-type tools on library projects. This is a general limitation, not just for EF.

**Comment by piotrpMSFT** [comment]
@weshaggard @ericstj @davidfowl @yishaigalatzer  for their thoughts on this one. Maybe we can have the library target both netstandard1.x and netcoreapp1.0. The tool would run against the netcoreapp target while we would only ship the netstandard target. Fairly easy to do via command line, but do we want to make this a cleaner experience? What are other options?

**Comment by natemcmaster** [comment]
Could work, but I've found this buggy. For example, this project.json has the same error as the original issue above. No runtimeconfig.json is produced for netcoreapp1.0.

```json
{
  "version": "1.0.0-*",
  "dependencies": {
    "Microsoft.EntityFrameworkCore.Sqlite": "1.0.0-*"
  },
  "frameworks": {
    "netstandard1.5": {
      "imports": [
        "portable-net452+win81"
      ],
      "dependencies": {
        "NETStandard.Library": "1.5.0-*"
      }
    },
    "netcoreapp1.0": { 
      "compilationOptions": {
        "emitEntryPoint": true
      },
      "imports": [
        "portable-net452+win81"
      ],
      "dependencies": {
        "Microsoft.NetCore.App": {
          "type": "platform",
          "version": "1.0.0-*"
        }
      }
    }
  }
}
```

**Comment by ericstj** [comment]
Which is the run-able entity here, the TOOL?  In that case should the restore be happening for the TOOL's TFM and merging the dependencies of the project as if it were a dependency of the tool?

**Comment by piotrpMSFT** [comment]
I chatted with @ericstj today. He'll put some thought into the model and may have feedback/suggestions. For the time being, however, we need to understand why the project.json above is not producing a deps file/runnable output.

**Comment by eerhardt** [comment]
I pulled https://github.com/natemcmaster/dotnet-sample-tool and was able to get `Pass. Called inside the project. :)` printed out by build.ps1 for SampleLibrary.  **The thing I needed to fix is to put `netcoreapp1.0` as the first TFM in the project.json**.  The CLI is using the first TFM in the file to run.

I chatted with @brthor today, and he mentioned a decent idea: If the library isn't already generating the necessary runtimeconfig.json file, maybe the CLI should generate the runtime files for the library as if the project.json was targetting `netcoreapp1.0`. That way, it wouldn't be the responsibility of the library developer. Instead the tools would have the responsibility of making this scenario work.

Thoughts?

**Comment by eerhardt** [comment]
I've figured out why I needed to put `netcoreapp1.0` as the first TFM in the project.json - it was because of this line in the `dotnet-my-tool` code:

https://github.com/natemcmaster/dotnet-sample-tool/blob/master/dotnet-my-tool/Program.cs#L21

```C#
    var framework = project.GetTargetFrameworks().Select(f=>f.FrameworkName).First();
```

The tool was only running against the first TFM, which in this case was `netstandard1.5`, which isn't a "runnable" TFM.

I've pushed some changes to the `dotnet-my-tool` in my fork to try invoking for all TFMs until it finds one that works:  https://github.com/eerhardt/dotnet-sample-tool/commit/1f97aab0e21891ce0df625604fc550c1aaced038

The CLI behavior is currently by design. In order for dependency tools to be invoked, they need to be invoked on a runnable TFM - namely `netcoreapp`.  To run them against a library you would either need to:

1. Add `netcoreapp1.0` as a TFM to the project.json.
 - Note the tool will need to find this runnable TFM and use it.
2. Add a new project that references the library, and invoke the tool on it. (like @natemcmaster mentions above)

There isn't anything the CLI can do in this area in the RC2 timeframe. Pushing to RTM in the hopes we can get support for this scenario from CoreFX. Note: the general scenario is pretty common: you have a tool that wants to load up a library and start invoking code in it. Test runners have this exact same scenario as well. The problem is that you need a runtime to run the app, but there isn't a runtime for the library. And you can't use the tool's runtime because it may not be the correct version that the library needs.

**Comment by natemcmaster** [comment]
Has there been any progress on this? There has been a lot of chatter about this limitation, especially because the workarounds are not easy.

**Comment by Villason** [comment]
Potentially related to https://github.com/aspnet/EntityFramework/issues/5460

**Comment by kjbetz** [comment]
@natemcmaster I put a different solution for https://github.com/aspnet/EntityFramework/issues/5460  here: https://github.com/aspnet/EntityFramework/issues/5460#issuecomment-221167712

cc/ @Villason

**Comment by ahmedanis03** [comment]
Ok
there is really a big headache from this issue and i am stucked  now.
first i was have this issue  #https://github.com/aspnet/EntityFramework/issues/5460
then applied this work around https://docs.efproject.net/en/latest/cli/dotnet.html#dotnet-cli-issues
another issue raised cannot remember the link but it is related that buildoptions should be on root in project.json fixed that and Yes! dotnet ef working fine with the commands and project running.
Now i added a new entity in the "Converted Class library" updated in the database and start using it in the Web App project and it is building fine and  then trying to run on iis it gave me this error

`System.TypeLoadException
Could not load type 'ApplicationDataModel.Models.CarsAd' from assembly 'ApplicationDataModel, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'.`

What ! build again and again using visual studio and nothing 
then i build the "Converted Class Library" using dotnet build in the package manager console and it seems that the whole project cannot be discovered as it cannot read the dbcontext class
`System.TypeLoadException
Could not load type 'ApplicationDataModel.Models.ApplicationDBContext' from assembly 'ApplicationDataModel, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'.`

```
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.TypeLoadException: Could not load type 'ApplicationDataModel.Models.ApplicationDBContext' from assembly 'ApplicationDataModel, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'.
   at DubaiAlNasClassifiedAds.Startup.ConfigureServices(IServiceCollection services)
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at Microsoft.AspNetCore.Hosting.Startup.ConfigureServicesBuilder.Invoke(Object instance, IServiceCollection exportServices)
   at Microsoft.AspNetCore.Hosting.Startup.ConfigureServicesBuilder.<>c__DisplayClass4_0.b__0(IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.Internal.WebHost.EnsureApplicationServices()
   at Microsoft.AspNetCore.Hosting.Internal.WebHost.BuildApplication()
```
 and this is weird as i check the output and visual studio is using dotnet.exe !!! so why the assembly only updated when i used dotnet build via package manager.


my ApplicationDataModel Project.json:

```
{
  "version": "1.0.0-*",
  "description": "ApplicationDataModel Class Library",
  "authors": [ "Ahmed" ],
  "frameworks": {
    "net46": {
      "buildOptions": {
        "emitEntryPoint": true
      }
    }
  },
  "buildOptions": {
    "emitEntryPoint": true
  },
  "dependencies": {
    "Microsoft.EntityFrameworkCore": "1.0.0-rc2-final",
    "Microsoft.EntityFrameworkCore.SqlServer": "1.0.0-rc2-final",
    "Microsoft.AspNetCore.Identity": "1.0.0-rc2-final",
    "Microsoft.AspNetCore.Identity.EntityFrameworkCore": "1.0.0-rc2-final",
    "Microsoft.EntityFrameworkCore.Tools": "1.0.0-preview1-final",
    "ImageProcessor": "2.3.3"
  },

  "tools": {

    "Microsoft.EntityFrameworkCore.Tools": {
      "version": "1.0.0-preview1-final",
      "imports": [
        "portable-net45+win8+dnxcore50",
        "portable-net45+win8"
      ]
    }
  }
}
```

dotnet -- info
```
.NET Command Line Tools (1.0.0-preview1-002702)

Product Information:
 Version:     1.0.0-preview1-002702
 Commit Sha:  6cde21225e

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.10586
 OS Platform: Windows
 RID:         win10-x64
```
Sorry for English mistakes

**Comment by natemcmaster** [comment]
@ahmedanis03 can you report this issue to https://github.com/aspnet/EntityFramework instead? Your error is different from the issue this bug was opened to track.

**Comment by natemcmaster** [comment]
We've had a few conversations internally about this. The most recent thinking is that the ASP.NET tooling that needs to load and execute user code can use an assembly load context to load the user's library into the tool process, rather than launching a new process.

 Using a load context hasn't been proven to work yet, and it will likely require some tweaks to the build system to make sure a class library gets the right artifacts anyways.

**Comment by kierenj** [comment]
Glad I found this issue here, very useful.  Sorry for the annoying question, but is there a ballpark ETA on this one - just so I can make some plans this end? (An order of magnitude is good enough for me! - 1d/1w/1m/1q/1y)

**Comment by natemcmaster** [comment]
Closing. In aspnet, we refactored our tools a long time ago to not use ProjectDependenciesCommandFactory. Also, this doesn't seem like something we could ever fix in the CLI: netstandard libraries projects are not executable.

**Comment by brthor** [comment]
@kierenj 

I agree this is a useful scenario in many cases. 
This has been possible for some time with **glorious hacks** (emphasis on hacks). 
Looks like official support is not coming so I'll share here.

You need to copy a runtimeconfig file to the command directory (after you've found the command but before executing). Fsharp actually did this in preview 1 (though the scenario was slightly different). 

https://github.com/dotnet/cli/blob/rel/1.0.0-preview1/src/dotnet-compile-fsc/Program.cs#L321

Some commands come packaged with a runtimeconfig to avoid this issue but I will assume that's not the case here. In your case you don't have a runtimeconfig, so you need to generate one somehow, then copy it. 
`dotnet new console && dotnet build` in a temp directory is the easiest candidate here but with plenty of caveats (YMMV). Copy the runtimeconfig from the bin directory.

This is the entirety of a runtimeconfig json file for a console app. You really just need to make sure the versions match your library (for which you may need more glorious hacks). 

```
{
  "runtimeOptions": {
    "tfm": "netcoreapp2.0",
    "framework": {
      "name": "Microsoft.NETCore.App",
      "version": "2.0.0"
    }
  }
}
```

Then use DepsJsonCommandResolver to execute the command. Refer to Fsc above for a complete example here.

https://github.com/dotnet/cli/blob/master/src/Microsoft.DotNet.Cli.Utils/CommandResolution/DepsJsonCommandResolver.cs

In summary, you use the library's `deps.json` to find the target dll among it's dependencies, and generate runtimeconfig which instructs the cli to use the default package set in it's installation, avoiding `A fatal error was encountered. The library 'libhostpolicy.dylib' required to execute the application was not found`.
